turbine/src/retransmit_stage.rs

```diff
@@ -845,74 +846,109 @@ mod tests {
     }
 
     #[test]
-    fn test_already_received() {
-        let slot = 1;
-        let index = 5;
-        let version = 0x40;
+    fn test_shred_deduper() {
         let keypair = get_keypair();
-        let mut shred = Shred::new_from_data(
-            slot,
-            index,
-            0,
-            &[],
-            ShredFlags::LAST_SHRED_IN_SLOT,
-            0,
-            version,
-            0,
-        );
-        shred.sign(&keypair);
+        let entries = create_ticks(10, 1, Hash::new_unique());
+        let rsc = ReedSolomonCache::default();
+        let make_shreds_for_slot = |slot, parent, code_index| {
+            let shredder = Shredder::new(slot, parent, 1, 0).unwrap();
+            shredder.entries_to_shreds(
+                &keypair,
+                &entries,
+                true,
+                // chained_merkle_root
+                Some(Hash::new_from_array(rand::thread_rng().gen())),
+                0,
+                code_index,
+                true,
+                &rsc,
+                &mut ProcessShredsStats::default(),
+            )
+        };
+
         let mut rng = ChaChaRng::from_seed([0xa5; 32]);
         let shred_deduper = ShredDeduper::<2>::new(&mut rng, /*num_bits:*/ 640_007);
-        // unique shred for (1, 5) should pass
-        assert!(!shred_deduper.dedup(shred.id(), shred.payload(), MAX_DUPLICATE_COUNT));
-        // duplicate shred for (1, 5) blocked
-        assert!(shred_deduper.dedup(shred.id(), shred.payload(), MAX_DUPLICATE_COUNT));
-        let mut shred = Shred::new_from_data(
-            slot,
-            index,
-            2,
-            &[],
-            ShredFlags::LAST_SHRED_IN_SLOT,
-            0,
-            version,
-            0,
+
+        // make a set of shreds for slot 5 with parent slot 4
+        let (shreds_data_5_4, shreds_code_5_4) = make_shreds_for_slot(5, 4, 0);
+        // make a set of shreds for slot 5 with parent slot 3
+        let (shreds_data_5_3, _shreds_code_5_3) = make_shreds_for_slot(5, 3, 0);
+        // make a set of shreds for slot 5 with parent slot 8
+        let (shreds_data_5_2, _shreds_code_5_8) = make_shreds_for_slot(5, 2, 0);
```

Comment:
nit: `s/_shreds_code_5_8/_shreds_code_5_2/`

Reply:
yes. 5-8 as you have suggested could not work since it would be parented in slot that is in the future. Fixed in https://github.com/anza-xyz/agave/commit/f9bbaf896854a488555e98d80d2228f6246763f0